File: web/src/cvss.ts
export const metricOrder = ['AV', 'AC', 'AT', 'PR', 'UI', 'VC', 'VI', 'VA', 'SC', 'SI', 'SA'];

export const metricLabels: Record<string, string> = {
  AV: 'Attack Vector',
  AC: 'Attack Complexity',
  AT: 'Attack Requirements',
  PR: 'Privileges Required',
  UI: 'User Interaction',
  VC: 'Vulnerable Confidentiality',
  VI: 'Vulnerable Integrity',
  VA: 'Vulnerable Availability',
  SC: 'Subsequent Confidentiality',
  SI: 'Subsequent Integrity',
  SA: 'Subsequent Availability',
};

export const metricOptions: Record<string, Array<{ value: string; label: string }>> = {
  AV: [
    { value: 'N', label: 'Network' },
    { value: 'A', label: 'Adjacent' },
    { value: 'L', label: 'Local' },
    { value: 'P', label: 'Physical' },
  ],
  AC: [
    { value: 'L', label: 'Low' },
    { value: 'H', label: 'High' },
  ],
  AT: [
    { value: 'N', label: 'None' },
    { value: 'P', label: 'Present' },
  ],
  PR: [
    { value: 'N', label: 'None' },
    { value: 'L', label: 'Low' },
    { value: 'H', label: 'High' },
  ],
  UI: [
    { value: 'N', label: 'None' },
    { value: 'P', label: 'Passive' },
    { value: 'A', label: 'Active' },
  ],
  VC: impactOptions(),
  VI: impactOptions(),
  VA: impactOptions(),
  SC: impactOptions(),
  SI: impactOptions(),
  SA: impactOptions(),
};

export function defaultMetrics(): Record<string, string> {
  return { AV: 'N', AC: 'L', AT: 'N', PR: 'N', UI: 'N', VC: 'N', VI: 'N', VA: 'N', SC: 'N', SI: 'N', SA: 'N' };
}

export function vectorFromMetrics(metrics: Record<string, string>): string {
  return `CVSS:4.0/${metricOrder.map((key) => `${key}:${metrics[key] || 'N'}`).join('/')}`;
}

export function metricsFromVector(vector: string): Record<string, string> {
  const metrics = defaultMetrics();
  for (const part of vector.split('/')) {
    const [key, value] = part.split(':');
    if (key in metrics && value) metrics[key] = value;
  }
  return metrics;
}

function impactOptions() {
  return [
    { value: 'H', label: 'High' },
    { value: 'L', label: 'Low' },
    { value: 'N', label: 'None' },
  ];
}

File: internal/cvss/cvss.go
package cvss

import (
	"fmt"
	"strings"

	gocvss40 "github.com/pandatix/go-cvss/40"
)

var BaseMetricOrder = []string{"AV", "AC", "AT", "PR", "UI", "VC", "VI", "VA", "SC", "SI", "SA"}

var BaseMetricValues = map[string][]string{
	"AV": {"N", "A", "L", "P"},
	"AC": {"L", "H"},
	"AT": {"N", "P"},
	"PR": {"N", "L", "H"},
	"UI": {"N", "P", "A"},
	"VC": {"H", "L", "N"},
	"VI": {"H", "L", "N"},
	"VA": {"H", "L", "N"},
	"SC": {"H", "L", "N"},
	"SI": {"H", "L", "N"},
	"SA": {"H", "L", "N"},
}

var MetricNames = map[string]string{
	"AV": "Attack Vector",
	"AC": "Attack Complexity",
	"AT": "Attack Requirements",
	"PR": "Privileges Required",
	"UI": "User Interaction",
	"VC": "Vulnerable System Confidentiality",
	"VI": "Vulnerable System Integrity",
	"VA": "Vulnerable System Availability",
	"SC": "Subsequent System Confidentiality",
	"SI": "Subsequent System Integrity",
	"SA": "Subsequent System Availability",
}

type Result struct {
	Vector   string
	Score    float64
	Severity string
	Metrics  map[string]string
}

func FromMetrics(metrics map[string]string) (Result, error) {
	parts := []string{"CVSS:4.0"}
	for _, key := range BaseMetricOrder {
		value := strings.ToUpper(metrics[key])
		if value == "" {
			return Result{}, fmt.Errorf("missing CVSS metric %s", key)
		}
		if !validValue(key, value) {
			return Result{}, fmt.Errorf("invalid %s value %q", key, value)
		}
		parts = append(parts, key+":"+value)
	}
	return FromVector(strings.Join(parts, "/"))
}

func FromVector(vector string) (Result, error) {
	parsed, err := gocvss40.ParseVector(vector)
	if err != nil {
		return Result{}, err
	}
	score := parsed.Score()
	severity, err := gocvss40.Rating(score)
	if err != nil {
		return Result{}, err
	}
	metrics := make(map[string]string, len(BaseMetricOrder))
	for _, key := range BaseMetricOrder {
		value, err := parsed.Get(key)
		if err != nil {
			return Result{}, err
		}
		metrics[key] = value
	}
	return Result{Vector: parsed.Vector(), Score: score, Severity: severity, Metrics: metrics}, nil
}

func validValue(metric, value string) bool {
	for _, candidate := range BaseMetricValues[metric] {
		if candidate == value {
			return true
		}
	}
	return false
}
